File: experiments/embedded-v2/tools/count-sloc.mjs
#!/usr/bin/env node

// Deterministic physical SLOC accounting for embedded-v2 candidates.
// Usage: node count-sloc.mjs path/to/sloc-manifest.json
//        node count-sloc.mjs --self-test

import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import process from "node:process";

const SCHEMA_VERSION = 1;
const SUPPORTED_SUFFIXES = new Set([".rs", ".toml", ".wat", ".wit", ".json", ".yaml", ".yml"]);
const IGNORED_DIRECTORIES = new Set([".git", "target"]);
const COUNTED_ROLES = new Set(["production"]);
const VALID_ROLES = new Set(["production", "test", "generated", "lockfile"]);

function fail(message) {
  throw new Error(message);
}

function sha256(filePath) {
  return crypto.createHash("sha256").update(fs.readFileSync(filePath)).digest("hex");
}

function styleFor(filePath) {
  const suffix = path.extname(filePath);
  if (suffix === ".rs") return "rust";
  if (suffix === ".wat" || suffix === ".wit") return "wasm-text";
  if (suffix === ".toml" || suffix === ".yaml" || suffix === ".yml") return "hash";
  if (suffix === ".json") return "json";
  fail(`unsupported source type: ${filePath}`);
}

function stripComments(line, style, state) {
  if (style === "json") return line;

  let lineComment;
  let blockOpen = null;
  let blockClose = null;
  if (style === "rust") {
    lineComment = "//";
    blockOpen = "/*";
    blockClose = "*/";
  } else if (style === "wasm-text") {
    lineComment = ";;";
    blockOpen = "(;";
    blockClose = ";)";
  } else if (style === "hash") {
    lineComment = "#";
  } else {
    fail(`unknown comment style: ${style}`);
  }

  let output = "";
  let index = 0;
  let quote = null;
  let escaped = false;
  while (index < line.length) {
    if (state.blockDepth > 0) {
      if (blockOpen && line.startsWith(blockOpen, index)) {
        state.blockDepth += 1;
        index += blockOpen.length;
      } else if (blockClose && line.startsWith(blockClose, index)) {
        state.blockDepth -= 1;
        index += blockClose.length;
      } else {
        index += 1;
      }
      continue;
    }

    const character = line[index];
    if (quote !== null) {
      output += character;
      if (escaped) escaped = false;
      else if (character === "\\") escaped = true;
      else if (character === quote) quote = null;
      index += 1;
      continue;
    }

    if (character === '"' || (style === "hash" && character === "'")) {
      quote = character;
      output += character;
      index += 1;
      continue;
    }
    if (line.startsWith(lineComment, index)) break;
    if (blockOpen && line.startsWith(blockOpen, index)) {
      state.blockDepth += 1;
      index += blockOpen.length;
      continue;
    }
    output += character;
    index += 1;
  }
  return output;
}

function countFile(filePath) {
  const style = styleFor(filePath);
  const state = { blockDepth: 0 };
  let count = 0;
  for (const line of fs.readFileSync(filePath, "utf8").split(/\r?\n/)) {
    if (stripComments(line, style, state).trim()) count += 1;
  }
  if (state.blockDepth !== 0) fail(`unterminated block comment: ${filePath}`);
  return count;
}

function walk(root, current = root, found = new Set()) {
  for (const entry of fs.readdirSync(current, { withFileTypes: true })) {
    if (IGNORED_DIRECTORIES.has(entry.name)) continue;
    const fullPath = path.join(current, entry.name);
    if (entry.isDirectory()) {
      walk(root, fullPath, found);
    } else if (entry.isFile()) {
      const suffix = path.extname(entry.name);
      if (SUPPORTED_SUFFIXES.has(suffix) || entry.name === "Cargo.lock") {
        found.add(path.relative(root, fullPath).split(path.sep).join("/"));
      }
    }
  }
  return found;
}

function analyze(manifestPathValue) {
  const manifestPath = fs.realpathSync(manifestPathValue);
  const manifest = JSON.parse(fs.readFileSync(manifestPath, "utf8"));
  if (manifest.schema_version !== SCHEMA_VERSION) fail(`schema_version must be ${SCHEMA_VERSION}`);
  if (typeof manifest.candidate !== "string" || !manifest.candidate) fail("candidate must be a non-empty string");
  if (typeof manifest.root !== "string" || !manifest.root) fail("root must be a non-empty string");
  if (!Array.isArray(manifest.entries)) fail("entries must be an array");

  const root = fs.realpathSync(path.resolve(path.dirname(manifestPath), manifest.root));
  if (!fs.statSync(root).isDirectory()) fail(`candidate root is not a directory: ${root}`);
  const attributed = new Set();
  const rows = [];
  const categoryTotals = {};
  const roleTotals = Object.fromEntries([...VALID_ROLES].sort().map((role) => [role, 0]));

  for (const entry of manifest.entries) {
    if (typeof entry !== "object" || entry === null || Array.isArray(entry)) fail("each entry must be an object");
    if (typeof entry.path !== "string" || !entry.path) fail("entry.path must be a non-empty string");
    const relative = entry.path.replaceAll("\\", "/");
    if (attributed.has(relative)) fail(`duplicate entry: ${relative}`);
    attributed.add(relative);
    if (!VALID_ROLES.has(entry.role)) fail(`invalid role for ${relative}: ${entry.role}`);
    if (typeof entry.category !== "string" || !entry.category) fail(`entry.category must be non-empty for ${relative}`);
    if ((entry.role === "generated" || entry.role === "lockfile") && !entry.reason) {
      fail(`${entry.role} entry requires a reason: ${relative}`);
    }

    const source = path.resolve(root, relative);
    const relativeCheck = path.relative(root, source);
    if (relativeCheck.startsWith("..") || path.isAbsolute(relativeCheck)) fail(`entry escapes candidate root: ${relative}`);
    if (!fs.existsSync(source) || !fs.statSync(source).isFile()) fail(`attributed file does not exist: ${relative}`);

    let sloc = 0;
    if (entry.role !== "generated" && entry.role !== "lockfile") sloc = countFile(source);
    roleTotals[entry.role] += sloc;
    if (COUNTED_ROLES.has(entry.role)) {
      categoryTotals[entry.category] = (categoryTotals[entry.category] ?? 0) + sloc;
    }
    rows.push({ path: relative, role: entry.role, category: entry.category, sloc, sha256: sha256(source) });
  }

  const discovered = walk(root);
  const missing = [...discovered].filter((item) => !attributed.has(item)).sort();
  const extra = [...attributed].filter((item) => !discovered.has(item)).sort();
  if (missing.length) fail(`unattributed supported files: ${JSON.stringify(missing)}`);
  if (extra.length) fail(`entries are not supported source/config files: ${JSON.stringify(extra)}`);

  rows.sort((left, right) => left.path.localeCompare(right.path));
  return {
    schema_version: SCHEMA_VERSION,
    candidate: manifest.candidate,
    manifest: manifestPath.split(path.sep).join("/"),
    manifest_sha256: sha256(manifestPath),
    counted_roles: [...COUNTED_ROLES].sort(),
    production_sloc: roleTotals.production,
    role_totals: roleTotals,
    category_totals: Object.fromEntries(Object.entries(categoryTotals).sort()),
    files: rows,
  };
}

function selfTest() {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), "embedded-v2-sloc-"));
  try {
    fs.mkdirSync(path.join(root, "src"));
    fs.writeFileSync(
      path.join(root, "src", "main.rs"),
      '// comment\nfn main() { /* inline */ }\n/* block\ncomment */\nlet x = "//";\n',
    );
    fs.writeFileSync(path.join(root, "Cargo.toml"), '# comment\n[package]\nname = "fixture#name"\n');
    const manifest = {
      schema_version: 1,
      candidate: "self-test",
      root: ".",
      entries: [
        { path: "Cargo.toml", role: "production", category: "configuration" },
        { path: "src/main.rs", role: "production", category: "host" },
      ],
    };
    const manifestPath = path.join(root, "sloc-manifest.json");
    fs.writeFileSync(manifestPath, JSON.stringify(manifest));
    const result = analyze(manifestPath);
    if (result.production_sloc !== 5) fail(`expected 5 SLOC, got ${result.production_sloc}`);
  } finally {
    fs.rmSync(root, { recursive: true, force: true });
  }
}

try {
  if (process.argv.length === 3 && process.argv[2] === "--self-test") {
    selfTest();
    console.log("self-test: ok");
  } else if (process.argv.length === 3) {
    console.log(JSON.stringify(analyze(process.argv[2]), null, 2));
  } else {
    fail("usage: node count-sloc.mjs <manifest.json> | --self-test");
  }
} catch (error) {
  console.error(`count-sloc: ${error.message}`);
  process.exitCode = 2;
}

File: experiments/embedded-v2/tools/count_sloc.py
#!/usr/bin/env python3
"""Deterministic physical SLOC accounting for embedded-v2 candidates.

Usage:
    python count_sloc.py path/to/sloc-manifest.json
    python count_sloc.py --self-test

The manifest is deliberately explicit: every supported source/configuration file
under the candidate root must be attributed.  That makes moving correctness
logic into an unusual directory visible instead of silently excluding it.
"""

from __future__ import annotations

import argparse
import hashlib
import json
import sys
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Any


SCHEMA_VERSION = 1
SUPPORTED_SUFFIXES = {".rs", ".toml", ".wat", ".wit", ".json", ".yaml", ".yml"}
IGNORED_DIRECTORIES = {".git", "target"}
COUNTED_ROLES = {"production"}
VALID_ROLES = COUNTED_ROLES | {"test", "generated", "lockfile"}


@dataclass
class CommentState:
    block_depth: int = 0


def sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def style_for(path: Path) -> str:
    if path.suffix == ".rs":
        return "rust"
    if path.suffix in {".wat", ".wit"}:
        return "wasm-text"
    if path.suffix in {".toml", ".yaml", ".yml"}:
        return "hash"
    if path.suffix == ".json":
        return "json"
    raise ValueError(f"unsupported source type: {path}")


def strip_comments(line: str, style: str, state: CommentState) -> str:
    if style == "json":
        return line

    if style == "rust":
        line_comment, block_open, block_close = "//", "/*", "*/"
    elif style == "wasm-text":
        line_comment, block_open, block_close = ";;", "(;", ";)"
    elif style == "hash":
        line_comment, block_open, block_close = "#", None, None
    else:
        raise ValueError(f"unknown comment style: {style}")

    output: list[str] = []
    index = 0
    quote: str | None = None
    escaped = False

    while index < len(line):
        if state.block_depth:
            if block_open and line.startswith(block_open, index):
                state.block_depth += 1
                index += len(block_open)
            elif block_close and line.startswith(block_close, index):
                state.block_depth -= 1
                index += len(block_close)
            else:
                index += 1
            continue

        character = line[index]
        if quote is not None:
            output.append(character)
            if escaped:
                escaped = False
            elif character == "\\":
                escaped = True
            elif character == quote:
                quote = None
            index += 1
            continue

        if character == '"' or (style == "hash" and character == "'"):
            quote = character
            output.append(character)
            index += 1
            continue

        if line.startswith(line_comment, index):
            break
        if block_open and line.startswith(block_open, index):
            state.block_depth += 1
            index += len(block_open)
            continue

        output.append(character)
        index += 1

    return "".join(output)


def count_file(path: Path) -> int:
    style = style_for(path)
    state = CommentState()
    count = 0
    with path.open("r", encoding="utf-8") as handle:
        for raw_line in handle:
            if strip_comments(raw_line, style, state).strip():
                count += 1
    if state.block_depth:
        raise ValueError(f"unterminated block comment: {path}")
    return count


def supported_files(root: Path) -> set[str]:
    discovered: set[str] = set()
    for path in root.rglob("*"):
        if not path.is_file():
            continue
        relative = path.relative_to(root)
        if any(part in IGNORED_DIRECTORIES for part in relative.parts):
            continue
        if path.suffix in SUPPORTED_SUFFIXES or path.name == "Cargo.lock":
            discovered.add(relative.as_posix())
    return discovered


def load_manifest(path: Path) -> dict[str, Any]:
    with path.open("r", encoding="utf-8") as handle:
        value = json.load(handle)
    if not isinstance(value, dict):
        raise ValueError("manifest root must be an object")
    return value


def analyze(manifest_path: Path) -> dict[str, Any]:
    manifest_path = manifest_path.resolve()
    manifest = load_manifest(manifest_path)
    if manifest.get("schema_version") != SCHEMA_VERSION:
        raise ValueError(f"schema_version must be {SCHEMA_VERSION}")
    candidate = manifest.get("candidate")
    if not isinstance(candidate, str) or not candidate:
        raise ValueError("candidate must be a non-empty string")
    root_value = manifest.get("root")
    if not isinstance(root_value, str) or not root_value:
        raise ValueError("root must be a non-empty string")
    root = (manifest_path.parent / root_value).resolve()
    if not root.is_dir():
        raise ValueError(f"candidate root is not a directory: {root}")

    raw_entries = manifest.get("entries")
    if not isinstance(raw_entries, list):
        raise ValueError("entries must be an array")

    attributed: set[str] = set()
    rows: list[dict[str, Any]] = []
    category_totals: dict[str, int] = {}
    role_totals: dict[str, int] = {role: 0 for role in sorted(VALID_ROLES)}

    for entry in raw_entries:
        if not isinstance(entry, dict):
            raise ValueError("each entry must be an object")
        relative_value = entry.get("path")
        role = entry.get("role")
        category = entry.get("category")
        if not isinstance(relative_value, str) or not relative_value:
            raise ValueError("entry.path must be a non-empty string")
        relative = Path(relative_value).as_posix()
        if relative in attributed:
            raise ValueError(f"duplicate entry: {relative}")
        attributed.add(relative)
        if role not in VALID_ROLES:
            raise ValueError(f"invalid role for {relative}: {role!r}")
        if not isinstance(category, str) or not category:
            raise ValueError(f"entry.category must be non-empty for {relative}")
        if role in {"generated", "lockfile"} and not entry.get("reason"):
            raise ValueError(f"{role} entry requires a reason: {relative}")

        source = (root / relative).resolve()
        try:
            source.relative_to(root)
        except ValueError as error:
            raise ValueError(f"entry escapes candidate root: {relative}") from error
        if not source.is_file():
            raise ValueError(f"attributed file does not exist: {relative}")

        lines = 0
        if role not in {"generated", "lockfile"}:
            lines = count_file(source)
        role_totals[role] += lines
        if role in COUNTED_ROLES:
            category_totals[category] = category_totals.get(category, 0) + lines
        rows.append(
            {
                "path": relative,
                "role": role,
                "category": category,
                "sloc": lines,
                "sha256": sha256(source),
            }
        )

    discovered = supported_files(root)
    missing = sorted(discovered - attributed)
    extra = sorted(attributed - discovered)
    if missing:
        raise ValueError(f"unattributed supported files: {missing}")
    if extra:
        raise ValueError(f"entries are not supported source/config files: {extra}")

    rows.sort(key=lambda row: row["path"])
    return {
        "schema_version": SCHEMA_VERSION,
        "candidate": candidate,
        "manifest": manifest_path.as_posix(),
        "manifest_sha256": sha256(manifest_path),
        "counted_roles": sorted(COUNTED_ROLES),
        "production_sloc": role_totals["production"],
        "role_totals": role_totals,
        "category_totals": dict(sorted(category_totals.items())),
        "files": rows,
    }


def self_test() -> None:
    with tempfile.TemporaryDirectory() as temporary:
        root = Path(temporary)
        (root / "src").mkdir()
        (root / "src" / "main.rs").write_text(
            "// comment\nfn main() { /* inline */ }\n/* block\ncomment */\nlet x = \"//\";\n",
            encoding="utf-8",
        )
        (root / "Cargo.toml").write_text(
            "# comment\n[package]\nname = \"fixture#name\"\n",
            encoding="utf-8",
        )
        manifest = {
            "schema_version": 1,
            "candidate": "self-test",
            "root": ".",
            "entries": [
                {"path": "Cargo.toml", "role": "production", "category": "configuration"},
                {"path": "src/main.rs", "role": "production", "category": "host"},
            ],
        }
        manifest_path = root / "sloc-manifest.json"
        manifest_path.write_text(json.dumps(manifest), encoding="utf-8")
        result = analyze(manifest_path)
        if result["production_sloc"] != 5:
            raise AssertionError(f"expected 5 SLOC, got {result['production_sloc']}")


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("manifest", nargs="?", type=Path)
    parser.add_argument("--self-test", action="store_true")
    arguments = parser.parse_args()
    try:
        if arguments.self_test:
            self_test()
            print("self-test: ok")
            return 0
        if arguments.manifest is None:
            parser.error("manifest is required unless --self-test is used")
        print(json.dumps(analyze(arguments.manifest), indent=2, sort_keys=True))
        return 0
    except (OSError, ValueError, json.JSONDecodeError, AssertionError) as error:
        print(f"count_sloc: {error}", file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())
